packages/zoltan2/src/algorithms/color/Zoltan2_Distance1_2GhostLayer.hpp

```diff
@@ -0,0 +1,521 @@
+                       Teuchos::RCP<femv_t> femv,
+                       bool recolor);
+    
+    RCP<const base_adapter_t> adapter;
+    RCP<GraphModel<base_adapter_t> > model;
+    RCP<Teuchos::ParameterList> pl;
+    RCP<Environment> env;
+    RCP<const Teuchos::Comm<int> > comm;
+    int numColors;
+
+  public:
+    AlgDistance1TwoGhostLayer(
+      const RCP<const base_adapter_t> &adapter_,
+      const RCP<Teuchos::ParameterList> &pl_,
+      const RCP<Environment> &env_,
+      const RCP<const Teuchos::Comm<int> > &comm_)
+    : adapter(adapter_), pl(pl_), env(env_), comm(comm_){
+
+      numColors = 4;
+      modelFlag_t flags;
+      flags.reset();
+      buildModel(flags);
+    }
+    
+
+    void constructSecondGhostLayer(std::vector<gno_t>& ownedPlusGhosts, //this argument changes
+                                   const std::vector<int>& owners,
+                                   ArrayView<const gno_t> adjs,
+                                   ArrayView<const offset_t> offsets,
+                                   RCP<const map_t> mapOwned,
+                                   std::vector< gno_t>& adjs_2GL,
+                                   std::vector< offset_t>& offsets_2GL) {
+      std::vector<int> sendcounts(comm->getSize(),0);
+      std::vector<gno_t> sdispls(comm->getSize()+1,0);
+      //loop through owners, count how many vertices we'll send to each processor
+      for(size_t i = 0; i < owners.size(); i++){
+        if(owners[i] != comm->getRank()&& owners[i] !=-1) sendcounts[owners[i]]++;
+      }
+      //construct sdispls (for building sendbuf), and sum the total sendcount
+      gno_t sendcount = 0;
+      for(int i = 1; i < comm->getSize()+1; i++){
+        sdispls[i] = sdispls[i-1] + sendcounts[i-1];
+        sendcount += sendcounts[i-1];
+      }
+      std::vector<gno_t> idx(comm->getSize(),0);
+      for(int i = 0; i < comm->getSize(); i++){
+        idx[i]=sdispls[i];
+      }
+      //construct sendbuf to send GIDs to owning processes
+      std::vector<gno_t> sendbuf(sendcount,0);
+      for(size_t i = offsets.size()-1; i < owners.size(); i++){
+        if(owners[i] != comm->getRank() && owners[i] != -1){
+          sendbuf[idx[owners[i]]++] = ownedPlusGhosts[i];
+        }
+      }
+      //remap the GIDs so we receive the adjacencies in the same order as the current processes LIDs
+      for(gno_t i = 0; i < sendcount; i++){
+        ownedPlusGhosts[i+offsets.size()-1] = sendbuf[i];
+      }
+      
+      //communicate GIDs to owners
+      Teuchos::ArrayView<int> sendcounts_view = Teuchos::arrayViewFromVector(sendcounts);
+      Teuchos::ArrayView<gno_t> sendbuf_view = Teuchos::arrayViewFromVector(sendbuf);
+      Teuchos::ArrayRCP<gno_t>  recvbuf;
+      std::vector<int> recvcounts(comm->getSize(),0);
+      Teuchos::ArrayView<int> recvcounts_view = Teuchos::arrayViewFromVector(recvcounts);
+      Zoltan2::AlltoAllv<gno_t>(*comm, *env, sendbuf_view, sendcounts_view, recvbuf, recvcounts_view);
```

Comment:
Is this a slow reimplementation of Distributor?  Epetra and Tpetra have plenty of functions for doing things like "figuring out what process owns what GIDs."  `Zoltan2::AlltoAllv` will have every process send a message to every other process (!!!).

Reply:
The purpose of this AlltoAllv is to figure out which owned vertices are ghosted on which processes. The ghost owners send ghosted GIDs back to their owners (which I determine the Tpetra way). The owners then send  degree and adjacency information back to the ghost owners. If there is a more efficient way of determining where owned vertices are ghosted from the owning process, that would be preferable over this.

Reply:
@ibogle You can do this with Import and Export on a CrsGraph whose rows are the GIDs, and whose (global) column indices are MPI process ranks. 

1. Construct a Map, `rowMapAll`, whose indices are all the GIDs (owned and ghosted) on each process.
2. Construct a Map, `rowMapOwned`, whose indices are the owned GIDs on each process.
3. Construct an Import object: `Import<...> importer(rowMapOwned, rowMapAll);`.
4. Construct a CrsGraph, `graphOwned`, whose row Map is `rowMapOwned`.
5. On each process p, insert p as a global column index once into each row of `graphOwned`.
6. Construct a CrsGraph, `graphAll`, whose row Map is `rowMapAll`.
7. Import to populate `graphAll`: `graphAll.doImport(graphOwned, importer, INSERT);`

Now global row `gid` of `graphAll` has as its (global) column indices, all the ranks of the MPI processes that either ghost or own `gid`.  At this point, you can populate other data structures and do a reverse-mode Export (with `importer`) to send information back to the owners.

Reply:
@ibogle Let's discuss this suggestion at our meeting next week before you dive in to this change.